docs/advanced/optimizing-selectors.md

```diff
@@ -0,0 +1,112 @@
+
+## Memoization
+
+Selectors are memoized functions. Memoized functions are calculated when their arguments change and the results are cached. Regardless of how many components or services consume a selector, a selector will calculate only once when state changes and the cached result will be returned to all consumers. Taking advantage of this feature can result in performance increases.
+
+For example, there exists this state model:
+
+```ts
+interface SomeStateModel {
+  data: Data[];
+  name: string;
+}
+```
+
+And in this example there is an input component where a user can type a name. On key down, an action is dispatched updating the `name` property of state. On the same page, another component renders `data`. In order to render state data we create a selector in our state class:
+
+```ts
+@Selector()
+static getViewData(state: SomeStateModel) {
+   return state.data.map(d => expensiveFunction(d));
+}
+```
+
+Selectors defined in state classes implicitly have `state` injected as their first argument. The above selector will be recalculated every time the user types into the input component. Since `state` could update rapidly when a user types, the expensive selector will needlessly recalculate even though it does not care about the `name` property of `state` changing. This selector does not take advantage of memoization.
+
+One way to solve this problem is to turn off the `injectContainerState` selector [option](options.md). This option does not apply to selectors declared outside of state classes. This option only applies to selectors declared within state classes by preventing state being implicitly injected as the first argument. This requires you to explicitly specify all arguments using the `@Selector([...])` decorator _to selectors that are declared in state classes_. For example, we create two selectors in our state class:
+
+```ts
+@Selector([SomeState])
+static getData(state: SomeStateModel) {
+   return state.data;
+}
+
+@Selector([SomeState.getData])
+static getViewData(data: Data[]) {
+   return data.map(d => expensiveFunction(d));
+}
+```
+
+This `getViewData` selector will not be recalculated when a user types into the input component. This selector targets the specific property of `state` it cares about as its argument by leveraging an additional selector. When the `name` property of state changes, the `getViewData` arguments _do not change_. Memoization is taken advantage of.
+
+An alternative solution to turning off the selector option is to create a [meta selector](../concepts/select#meta-selectors). For example, we declare one selector in our state class and declare another selector outside of our state class:
+
+```ts
+@State({...})
+@Injectable()
+export class SomeState {
+  @Selector()
+  static getData(state: SomeStateModel) {
+    return state.data;
+  }
+}
+
+@Injectable()
+export class SomeSelectorService {
```

Comment:
I wouldn't call it service and it's definitely shouldn't have the injectable decorator since only static methods will be used. I remember we had examples in our docs with a class that has exposed selectors and its name was suffixed with "Selectors". I would like just to keep it consistent :slightly_smiling_face:

Reply:
Excellent catch 🎣  Thank you!

I will remove `@Injectable()` and rename this class to `SomeSelectors` to remain consistent with the other class names in that example.

Reply:
PS. I actually prefer to name these classes with the `Queries` suffix.
For example `TodoQueries`